File: crates/aura-hooks/plugins/aura-agent/extensions/pi.ts
// Aura's Pi extension: report what the agent changed, so the repo shows as
// worked-in rather than idle.
//
// Like OpenCode, Pi has no shell-command hooks — an extension is a module it
// imports — so this is a shim, not a decision. What counts as a change lives
// in `on-post-tool-use.sh` next door, the one body every non-Claude CLI runs.
//
// Staged by `aura-hooks` at `~/.pi/agent/extensions/aura.ts`, where Pi
// discovers it for every project. Nothing is written into the repo.
//
// Pi splits a tool call across two events and only the first carries the
// arguments, so the arguments are held until the call finishes — an edit that
// was blocked or errored should not be recorded as one that happened.

import { spawn } from "node:child_process";
import { homedir } from "node:os";
import { join } from "node:path";

const HOOK = join(homedir(), ".aura", "plugins", "aura-agent", "scripts", "on-post-tool-use.sh");

/** Hand one tool call to the shared hook and forget about it. Failure is not
 *  worth surfacing: the agent's own work succeeded, and only the record of it
 *  was missed. */
const report = (payload: Record<string, unknown>) => {
  try {
    const child = spawn(HOOK, {
      cwd: process.cwd(),
      env: { ...process.env, AURA_HOOK_AGENT: "Pi" },
      stdio: ["pipe", "ignore", "ignore"],
    });
    child.on("error", () => {});
    child.stdin?.on("error", () => {});
    child.stdin?.end(JSON.stringify(payload));
  } catch {
    // no-op
  }
};

/** What Pi hands a handler as its second argument. Only the one method is
 *  named, because it is the only one this extension has any business calling:
 *  Pi is alone among these CLIs in not putting the session id on the event, so
 *  it has to be asked for. */
type PiContext = { sessionManager?: { getSessionId?: () => string } };

export default function aura(pi: {
  on: (
    event: string,
    handler: (event: Record<string, any>, ctx: PiContext) => Promise<void> | void,
  ) => void;
}) {
  const pending = new Map<string, { toolName: string; args: unknown }>();

  pi.on("tool_execution_start", (event) => {
    if (typeof event.toolCallId === "string") {
      pending.set(event.toolCallId, { toolName: event.toolName, args: event.args });
    }
  });

  pi.on("tool_execution_end", (event, ctx) => {
    const id = typeof event.toolCallId === "string" ? event.toolCallId : "";
    const started = pending.get(id);
    pending.delete(id);
    if (!started || event.isError === true) return;

    // Pi identifies a session by its session file rather than putting an id on
    // the event, so the id comes off the context. Absent is survivable — the
    // row is still worth writing, it just cannot be grouped.
    let sessionId: string | undefined;
    try {
      sessionId = ctx?.sessionManager?.getSessionId?.();
    } catch {
      // no-op
    }

    report({
      toolName: started.toolName,
      args: started.args,
      sessionId,
      cwd: process.cwd(),
    });
  });
}

File: crates/aura-hooks/src/pi.rs
//! Pi's extension.
//!
//! Like OpenCode, Pi loads modules rather than running shell commands, and it
//! discovers every file in `~/.pi/agent/extensions/` for every project —
//! checked here with a marker file, not read off a page. So Aura's extension
//! is a shim over the shared script.
//!
//! The one thing Pi does differently: a tool call arrives as two events, and
//! only the first carries the arguments. The extension holds them until the
//! call finishes, so a tool that was blocked or errored is not recorded as an
//! edit that happened.
//!
//! The file is `aura.ts`, and the name matters more than it looks: this
//! directory is shared, and at least one other tool stamps its own extension
//! beside ours.

use std::path::PathBuf;

/// The extension body, compiled in.
pub const AURA_PI_EXTENSION: &str = include_str!("../plugins/aura-agent/extensions/pi.ts");

/// Stage the shared script and drop the extension into Pi's global extension
/// directory.
pub fn stamp_pi_extension() -> Option<()> {
    crate::shared::stage_shared_scripts()?;
    let mut path = pi_agent_dir()?;
    path.push("extensions");
    path.push("aura.ts");
    crate::opencode::write_if_changed(&path, AURA_PI_EXTENSION)
}

/// `~/.pi/agent`, and only if it is already there — Pi creates it on first
/// run, so a missing one means Pi has never run on this machine.
fn pi_agent_dir() -> Option<PathBuf> {
    let mut p = PathBuf::from(std::env::var_os("HOME")?);
    p.push(".pi");
    p.push("agent");
    p.is_dir().then_some(p)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn the_extension_pipes_into_the_shared_script_rather_than_deciding_for_itself() {
        assert!(AURA_PI_EXTENSION.contains("on-post-tool-use.sh"));
        assert!(
            !AURA_PI_EXTENSION.contains("aura log-intent"),
            "the shim has started making the decision itself",
        );
    }

    #[test]
    fn it_waits_for_the_call_to_finish_before_recording_it() {
        // Pi splits a tool call in two and only the start carries arguments.
        // Recording on the start event would log edits that were then blocked
        // by a permission prompt or failed outright.
        assert!(AURA_PI_EXTENSION.contains("tool_execution_start"));
        assert!(AURA_PI_EXTENSION.contains("tool_execution_end"));
        let start = AURA_PI_EXTENSION.find("tool_execution_start").unwrap();
        let end = AURA_PI_EXTENSION.find("tool_execution_end").unwrap();
        let report = AURA_PI_EXTENSION.rfind("report({").expect("it reports somewhere");
        assert!(report > end && end > start, "the report moved off the end event");
        assert!(AURA_PI_EXTENSION.contains("isError"), "a failed call is recorded as an edit");
    }

    #[test]
    fn it_names_itself_so_the_intent_row_says_who_was_working() {
        assert!(AURA_PI_EXTENSION.contains("AURA_HOOK_AGENT"));
        assert!(AURA_PI_EXTENSION.contains("\"Pi\""));
    }

    #[test]
    fn it_asks_for_the_session_id_rather_than_reading_one_off_the_event() {
        // Pi is the only one of these CLIs that does not put a session id on
        // the tool event; it identifies a session by its file, reachable only
        // through the handler's context.
        assert!(AURA_PI_EXTENSION.contains("getSessionId"));
        assert!(AURA_PI_EXTENSION.contains("sessionId"));
    }
}
